### DISCUSSION TITLE
Can we produce dependentRequired or other conditions somehow?

### DISCUSSION THREAD

**Discussion by HenrikPublic** [victools/jsonschema-generator#328] [Q&A] [answered]
Is there any way we can annotate a class in some magic way to represent conditions and have the generator produce such schemas?

Look at the examples here: https://json-schema.org/understanding-json-schema/reference/conditionals.html
How would we describe this in a way that would make the generator produce the correct conditions?

**Comment by CarstenWickner** [comment]
Hi @HenrikPublic,

The short answer is: No, no magic annotation (that I know of) would fully cover the nearly endless possibilities there.
The main challenge is the `if`/`then`/`else` subschemas, which often don't correspond to a Java type, i.e., need to be "hand-crafted". The same goes for `dependentSchemas`.

If you have a concrete example, I may be able to provide you with a concrete suggestion for that use case.

**Comment by HenrikPublic** [comment]
Thought soo, unfortunately 😊 

I was dreaming about something like
```
@NotNull Value required;
Value optional;
@RequiredIf("optional") OtherValue dependentRequired;
```

For if-then-else, I am not sure how I envisioned this. Perhaps with a list of annotations:
```
@NotNull Country country;
@IfThenElse {
  @MagicCheck("country",Country.US
  @Pattern(regexp = "some-pattern"),
  @Pattern(regexp = "some-other-pattern)
} PostalCode postalCode
```

The allOf and anyOf probably gets even more contrived...

I just hoped someone already saw that need and had thought of a clever solution ;-)

**Reply by CarstenWickner** [reply]
I was considering adding the `dependentRequired` keyword, but was unsure what the more intuitive way to configure it:
1. On a given property, ask which other properties are required if this one is present.
2. On a given property, ask which other property should be present to make this one required.

Option 1 would be close to the resulting schema:
> In the following example, whenever a `credit_card` property is provided, a `billing_address` property must also be present:
> ```json
> {
>   "dependentRequired": {
>     "credit_card": ["billing_address"]
>   }
> }
> ```
In your example with the `@RequiredIf` annotation, you went for option 2.

Coming up with an annotation (either `@IfPresentAlsoRequire({"billing_address"})` or `@RequiredIf("credit_card")` doesn't matter too much) is simple enough – I merely need to provide a corresponding configuration option.
For that, I just need to find a name as well. 😅
What do you think about the following?
```java
SchemaGeneratorConfigPart<M> withDependentRequiresResolver(ConfigFunction<M, List<String>> check);
```
For example:
```java
configBuilder.forFields().withDependentRequiresResolver(field -> Optional
        .ofNullable(field.getAnnotationConsideringFieldAndGetter(IfPresentAlsoRequire.class)
        .map(IfPresentAlsoRequire::value)
        .map(Arrays::asList)
        .orElse(null));
```
A `@RequiredIf` annotation could also be used, it'd just involve scanning all other fields in the declaring type, i.e., would look a little more complicated – but nevertheless is possible.

----

As for `if`/`then`/`else`, you can hand-craft schemas containing those keywords via a `CustomDefinitionProvider`. Due to the involved complexity, you should be able to create an annotation that matches your more specific requirements and then create a configuration based on that.
I have no good idea, how that could be handled in a meaningful flexible way by the schema generator directly.

**Reply by CarstenWickner** [reply]
Started working on `dependentRequired` now, but cannot yet promise when it's done: https://github.com/victools/jsonschema-generator/pull/330

**Reply by HenrikPublic** [reply]
Cool. Thx a lot. I am more than happy with option 1.

**Reply by HenrikPublic** [reply]
For the if/then/else with CustomDefinitionProvider and hand-made annotations, I probably need to study a bit more 😊 
But, what you are saying is that I need a hand-crafted one for each new if-then-else?

**Reply by CarstenWickner** [reply]
If you have a common pattern, one (or a few) options on your custom annotations might be sufficient.
You could also expect those conditions to be provided as complete schema strings, as a shortcut.

I can provide a few examples, once I'm done with the other changes and find some time. 😅

**Reply by HenrikPublic** [reply]
Examples would be much appreciated!

**Reply by CarstenWickner** [reply]
I've created an example that looks like this:
```java
    @SchemaCondition(
            ifFulFilled = @SchemaProperty(name = "country", constValue = "US"),
            thenExpect = @SchemaProperty(name = "postalCode", pattern = "^\\d{5}(?:[-\\s]\\d{4})?$"),
            elseExpect = @SchemaProperty(name = "postalCode", pattern = "^[-\\s\\d]+$")
    )
    static class TestType {

        String country;
        String postalCode;
    }

    @Target({ElementType.FIELD, ElementType.METHOD, ElementType.TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    @interface SchemaCondition {
        SchemaProperty ifFulFilled();
        SchemaProperty[] thenExpect() default {};
        SchemaProperty[] elseExpect() default {};
    }

    @Target({ElementType.ANNOTATION_TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    @interface SchemaProperty {
        String name();
        String constValue() default "";
        String pattern() default "";
    }
```
The full example is here:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/main/java/com/github/victools/jsonschema/examples/IfThenElseExample.java

The generated output would look like this:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/test/resources/com/github/victools/jsonschema/examples/IfThenElseExample-result.json

**Reply by HenrikPublic** [reply]
That's really helpful! Thanks a lot :-)

**Reply by CarstenWickner** [reply, marked as answer]
Release [v4.30.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.30.0) has been published. 🚀

It includes support for the `dependentRequired` keyword as per the aforementioned PR.
